File: src/sliding-window/sliding-window-maximum.ts
export function maxSlidingWindow(nums: number[], k: number): number[] {
  const queue: number[] = []
  const n = nums.length
  const ans: number[] = []
  for (let i = 0; i < n; ++i) {
    while (queue.length && queue[0] < i - k + 1) {
      queue.shift()
    }

    while (queue.length && nums[queue[queue.length - 1]] <= nums[i]) queue.pop()

    queue.push(i)

    if (i - k + 1 >= 0) {
      ans.push(nums[queue[0]])
    }
  }

  return ans
}

File: src/sliding-window/sliding-window-maximum.cpp
//单调队列解法
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        deque<int> q;
        vector<int> ans;
        for(int i = 0; i < n; ++i) {
            if(q.size() && i - q.front() >= k) q.pop_front();

            while(q.size() && nums[q.back()] <= nums[i]) q.pop_back();

            q.push_back(i);

            if(i - k + 1 >= 0) {
                ans.push_back(nums[q.front()]);
            }
        }

        return ans;
    }
};